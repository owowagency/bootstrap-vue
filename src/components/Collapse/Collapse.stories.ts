import Btn from '@/components/Button';
import Card from '@/components/Card';
import Collapse from '.';

export default {
    title: 'Components/Collapse',
    component: Collapse,
};

const template = (args) => ({
    components: {Btn, Card, Collapse},
    setup: () => ({args}),
    template: `
        <Collapse v-bind="args">
            <template #toggle-content="{id}">
                <Btn
                    class="mb-2"
                    data-bs-toggle="collapse"
                    :data-bs-target="'#' + id"
                >
                    Open me
                </Btn>
            </template>

            <Card text="Some placeholder content for the collapse component. This panel is hidden by default but revealed when the user activates the relevant trigger." />
        </Collapse>
    `,
});

export const Default = template.bind({});

