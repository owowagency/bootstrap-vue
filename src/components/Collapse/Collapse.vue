<template>
    <slot
        :id="id"
        name="toggle"
    >
        <div
            class="collapse-toggle"
            data-bs-toggle="collapse"
            :data-bs-target="`#${id}`"
            v-bind="$attrs"
        >
            <slot
                :id="id"
                name="toggleContent"
            />
        </div>
    </slot>

    <div
        :id="id"
        ref="collapse"
        class="collapse"
        :data-bs-parent="parentSelector"
    >
        <slot :id="id" />
    </div>
</template>

<script lang="ts">
import {idProps} from '@/composables/useId';
import {ref} from 'vue';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';

const collapseEvents = ['show', 'shown', 'hide', 'hidden'] as const;

type CollapseEvent = typeof collapseEvents[number];
</script>

<script lang="ts" setup>
defineProps({
    ...idProps,
    // Should be a valid query selector.
    parentSelector: {
        type: String,
        default: null,
    },
});

const emit = defineEmits<{(event: CollapseEvent): void}>();

const collapse = ref<HTMLElement>();

useBootstrapEmits(
    collapse,
    collapseEvents,
    emit,
    'collapse',
);

const {bsInstance: bsCollapse} = useBootstrapInstance(
    'Collapse',
    collapse,
    {toggle: false},
);

defineExpose({bsCollapse});
</script>
