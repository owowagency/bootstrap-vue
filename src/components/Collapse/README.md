# Collapse

```vue
<Collapse parent-selector="#parent">
    <template #toggle-content="{id}">
        <Button
            data-bs-toggle="collapse"
            :data-bs-target="'#' + id"
        >
            Open me
        </Button>
    </template>

    <Card text="This panel is hidden by default but revealed when the user activates the relevant trigger." />
</Collapse>
```

<!-- TODO Auto-generate based on type? -->

## Events

| Name   | Args | Description                                                                   |
| ------ | ---- | ----------------------------------------------------------------------------- |
| show   |      | This event fires immediately when the `show` instance method is called        |
| shown  |      | This event is fired when a collapse element has been made visible to the user |
| hide   |      | This event is fired immediately when the `hide` method has been called        |
| hidden |      | This event is fired when a collapse element has been hidden from the user     |

## Props

| Prop name      | Description                                                                    | Type    | Values | Default |
| -------------- | ------------------------------------------------------------------------------ | ------- | ------ | ------- |
| parentSelector | The selector of the parent element<br/>_This should be a valid query selector_ | string  | -      | null    |
| visible        | Whether the collapse is visible                                                | boolean | -      | false   |

## Slots

| Name          | Description                          | Bindings                                        |
| ------------- | ------------------------------------ | ----------------------------------------------- |
| toggle        | Displays the toggle for the collapse | **id** `string` - The id of the target collapse |
| toggleContent | Displays the content for the toggle  | **id** `string` - The id of the target collapse |
| default       | Displays the content of the collapse | **id** `string` - The id of the collapse        |
