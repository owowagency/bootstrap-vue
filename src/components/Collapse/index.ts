import Collapse from './Collapse.vue';

export default Collapse;
