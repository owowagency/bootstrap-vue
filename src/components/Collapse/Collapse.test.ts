import Collapse from '.';
import {shallowMount} from '@vue/test-utils';

jest.mock('bootstrap', () => ({
    Collapse: {
        getOrCreateInstance: jest.fn(),
    },
}));

describe('template', () => {
    componentRenderTest(Collapse, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentSlotRenderTest(Collapse, 'toggle');

    componentSlotRenderTest(Collapse, 'toggleContent');

    ['show', 'shown', 'hide', 'hidden'].forEach((event: string) => {
        componentBootstrapEventTest(
            Collapse,
            '.collapse',
            event,
            'collapse',
        );
    });
});

describe('onMounted', () => {
    it('sets bsCollapse', async() => {
        const wrapper = await shallowMount(Collapse);

        expect((await import('bootstrap')).Collapse.getOrCreateInstance).toBeCalledWith(wrapper.vm.$refs.collapse, {toggle: false});

        // TODO: Unable to assert bsCollapse value.
    });
});
