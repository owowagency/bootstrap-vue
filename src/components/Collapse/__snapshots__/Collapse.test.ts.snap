// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div class=\\"collapse-toggle\\" data-bs-toggle=\\"collapse\\" data-bs-target=\\"#dont-change-please\\"></div>
<div id=\\"dont-change-please\\" class=\\"collapse\\"></div>"
`;
